File: Javascript/35. Search Insert Position.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var searchInsert = function(nums, target) {
    let left = 0
        right = nums.length - 1
        mid = 0
    
    while (left <= right) {
        mid = Math.floor((left+right) / 2)

        if (nums[mid] === target) {
            return mid
        } else if (nums[mid] > target) {
            right = mid - 1
        } else {
            left = mid + 1
        }
    }

    return target > nums[mid] ? mid + 1 : mid
};

//  For testing
// const test = searchInsert(
//     [1,3,5,6],
//     4
// )
console.log(test)

File: Python/35. Search Insert Position.py
class Solution:
    def searchInsert(self, nums: list[int], target: int) -> int:
        left, right = 0, len(nums)-1

        while left <= right:
            mid = (left+right) // 2
            if nums[mid] == target:
                return mid
            elif target > nums[mid]:
                left = mid + 1
            else:
                right = mid - 1
        return mid + 1 if target > nums[mid] else mid
    


if __name__ == "__main__":
    solution = Solution()
    # For testing
    print(solution.searchInsert(
        nums=[1,3,5,6],
        target = 0
    ))